### ISSUE TITLE
Bad certificate with sidecar-injector webhook in 1.0.0-snapshot.2

### ISSUE THREAD

**Issue by liamawhite** [istio/istio#7478]
**Describe the bug**
Sidecar injector logs bad certificate when called by kube apiserver. So no pods will deploy.

```
istio-sidecar-injector-589646bb59-qvth6 sidecar-injector-webhook 2018-07-28T16:10:57.602516Z	info	http: TLS handshake error from 172.30.96.193:55162: remote error: tls: bad certificate
istio-sidecar-injector-589646bb59-qvth6 sidecar-injector-webhook 2018-07-28T16:10:57.949561Z	info	http: TLS handshake error from 172.30.96.193:55170: remote error: tls: bad certificate
istio-sidecar-injector-589646bb59-qvth6 sidecar-injector-webhook 2018-07-28T16:10:58.185100Z	info	http: TLS handshake error from 172.30.96.193:55178: remote error: tls: bad certificate
istio-sidecar-injector-589646bb59-qvth6 sidecar-injector-webhook 2018-07-28T16:10:58.493076Z	info	http: TLS handshake error from 172.30.96.193:55198: remote error: tls: bad certificate
istio-sidecar-injector-589646bb59-qvth6 sidecar-injector-webhook 2018-07-28T16:11:18.327897Z	info	http: TLS handshake error from 172.30.96.193:55846: remote error: tls: bad certificate
istio-sidecar-injector-589646bb59-qvth6 sidecar-injector-webhook 2018-07-28T16:11:20.938075Z	info	http: TLS handshake error from 172.30.96.193:55902: remote error: tls: bad certificate
```

```Liams-MBP:fundamentals liam$ k get MutatingWebhookConfiguration istio-sidecar-injector -o yaml
apiVersion: admissionregistration.k8s.io/v1beta1
kind: MutatingWebhookConfiguration
metadata:
  creationTimestamp: 2018-07-28T10:34:50Z
  generation: 2
  labels:
    app: istio-sidecar-injector
  name: istio-sidecar-injector
  resourceVersion: "36812"
  selfLink: /apis/admissionregistration.k8s.io/v1beta1/mutatingwebhookconfigurations/istio-sidecar-injector
  uid: db60ce6f-9251-11e8-b5b0-0e46067ded37
webhooks:
- clientConfig:
    caBundle: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUM1VENDQWMyZ0F3SUJBZ0lRRWdYaUg5VmVsYzVtUTJEY09CQjRwakFOQmdrcWhraUc5dzBCQVFzRkFEQWMKTVJvd0dBWURWUVFLRXhGck9ITXVZMngxYzNSbGNpNXNiMk5oYkRBZUZ3MHhPREEzTWpnd05qVTFNVEphRncweApPVEEzTWpnd05qVTFNVEphTUJ3eEdqQVlCZ05WQkFvVEVXczRjeTVqYkhWemRHVnlMbXh2WTJGc01JSUJJakFOCkJna3Foa2lHOXcwQkFRRUZBQU9DQVE4QU1JSUJDZ0tDQVFFQXN4NkJCSnFXclQzWXFrcERYUDdrdG5VYXpESW0KMnVmSjJIZ1JlV0ZwdzVrN0Y4NFV1TmdoYnNrdnlRZ2hpMkRlUXFHMzZrM0VoTkw0cGtGYU9ubXdIa1NFY2FNOAo4a1BGNXIrdTZRS2xvRVRwYzJBVytyUlRVRFdvS2dsdmhyL0IvUEVLN0FnZVQ3VXk1a3dTd2lBZ0w5OG1GVXVqCkhDUllNUGF5c2FtUVkxbXNreHhEK2tINS9YNHVNcU9DYVkrWDZDQnFtbkp3RmRUUENRcktaTVF2bjBDZXl6T3QKdHYvVWFSb1R3dWtNcnZvcFJmb2FoeE5uVlJXN0xaMWlhajZrWmhmRlR6SWpldTdPRDNUZElBbzUrc2h5Rm40SgpnUHYvaENQUmQ0aE9VSzhQRmZyTVlWQnZqNGxXN3N2VS9qcS9OTlpnMmxXMXdGZnhhd3FnTmhieUF3SURBUUFCCm95TXdJVEFPQmdOVkhROEJBZjhFQkFNQ0FnUXdEd1lEVlIwVEFRSC9CQVV3QXdFQi96QU5CZ2txaGtpRzl3MEIKQVFzRkFBT0NBUUVBbjU1S2hGTngzbE9EYmp4aTJVUVlCbjFRb3ZxdDN2ZUZjbWVTOFV2azMzaWxIS29ucVFHaQpZaHJPMEx5Vk1XYmV2V1h5OCsvMHlOVG8ySVozaWFyUUlaSk1qTkZUalhVRTY4b1N6N2ZqOW9XZUsyb2g2bE93CmdCUE9nVy9ZQ0FlK1g5RHRNVngrUUJMNlNyTjdWMmIrcVNWSEdXK0pOR3NoVFRsZzBHSW1QTDNMTEU2dU1vVWQKc0xZbFVxd2hlZExhY2VmVll6bFRxdHZNRy9XNk4yVkNDVVhvTS9BcUo4allYbXk3UlpnT1J1cEs2MVY5eWxHTQpWcmZuQ05PVnJMZWNlSVY1dWVCUG1KZkt0MUd0cnY4RmFoQlFxeWpyaVpQOXFNdWdFVC8vRzYzanlGeEdmbk5uCnVVVThlb2NaakxVQ0MxMFFzU0VNOURFaW9Ub3VpWmlORnc9PQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==
    service:
      name: istio-sidecar-injector
      namespace: istio-system
      path: /inject
  failurePolicy: Fail
  name: sidecar-injector.istio.io
  namespaceSelector:
    matchLabels:
      istio-injection: enabled
  rules:
  - apiGroups:
    - ""
    apiVersions:
    - v1
    operations:
    - CREATE
    resources:
    - pods
```
```Liams-MBP:fundamentals liam$ k get secret istio.istio-sidecar-injector -o yaml
apiVersion: v1
data:
  cert-chain.pem: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSURQekNDQWllZ0F3SUJBZ0lRWHFCdEdZODBqNnY0RFhwZVU2R3UvREFOQmdrcWhraUc5dzBCQVFzRkFEQWMKTVJvd0dBWURWUVFLRXhGck9ITXVZMngxYzNSbGNpNXNiMk5oYkRBZUZ3MHhPREEzTWpnd09UUTNORE5hRncweApPREV3TWpZd09UUTNORE5hTUFzeENUQUhCZ05WQkFvVEFEQ0NBU0l3RFFZSktvWklodmNOQVFFQkJRQURnZ0VQCkFEQ0NBUW9DZ2dFQkFQeHczSmVYQ3o2Z2xMZS9lQ0pDZytEdURwQ2MwTGFjZno0ekduZC8rYkErZXFpeVhiQzAKKzVtc2ZOK29lck45aDhMeTBOMEphSHN2dS9weUtncHlkSzBOOVVUV212QWkyT1g0eHBmelZ6QWRkRVMwdG8wQQpwbk1oTzJ3VDVNTTdNcWJWZ2JVbE1KOFEvQlFTdG81OEdUWEFzS1ROS2Z5VThvYVlCVnhneDd5cXlCOXl1dm5TCkgrdVNPWkpRTEtWd2FnN0JObEUvLzZwbUROdk5lcUptVy81NG5VTk9YM3d6QlZyZjM4ZTR5MERPTWEzeTl3QW0KNlFtR0ZtU2ZIdGFkSDNNU1dvMzVCcTh6L3J5bHc3cmJpNlhNTWQzWmZJZjVnQTlmeFNGdEhjTWVBZEhoVTRQMQo1MnlqWTR5RjlVV0pQV1hzOC8rU1BUSXJxU05vL3BxV3VFc0NBd0VBQWFPQmpUQ0JpakFPQmdOVkhROEJBZjhFCkJBTUNCYUF3SFFZRFZSMGxCQll3RkFZSUt3WUJCUVVIQXdFR0NDc0dBUVVGQndNQ01Bd0dBMVVkRXdFQi93UUMKTUFBd1N3WURWUjBSQkVRd1FvWkFjM0JwWm1abE9pOHZZMngxYzNSbGNpNXNiMk5oYkM5dWN5OXBjM1JwYnkxegplWE4wWlcwdmMyRXZhWE4wYVc4dGMybGtaV05oY2kxcGJtcGxZM1J2Y2pBTkJna3Foa2lHOXcwQkFRc0ZBQU9DCkFRRUFTMEhreTA4NVJ0VWQrdDZsR2RuRDM3bXovUThtekJUbnZsTzBBdS8zR0szY3dNRUtPV0Y1aC9RSmJyT2MKc3g0WE1wMk5IUHNTVFhwSXZtZWo4dUNCeTRCb0JqMGhNRFR3Z2NUWS9uK3hsMGtsT1dGempoL1RTRXdEQlplWQpvc25nK3MwVEdBMnF5UDJOdUl3UHpnSnFuUFZObTFZL3d0YVpQVnV2Mm51cFVrd2VyZ3liTUZ1Mm5CR2RTaG9uCjZtTDF6eEJtMHQ5SVBUZS9OdjZUdW8xTURBelkxWTg0MEpMcVBNUlZUWmVpZDNJb1Fxb2xuYjZmaGlKb3pOWC8KREdmd3JSbVdoWXIrdHpFdllzL3R5M2orb0tRbFBzUm1FeUlpS1ZZVktCWXlxRlhPd3QwSWc0eXdJa3hQQXAyYgo0Tzk5VUxsalYyWEg5a0JnWStZdmVtSkJLUT09Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K
  key.pem: LS0tLS1CRUdJTiBSU0EgUFJJVkFURSBLRVktLS0tLQpNSUlFcFFJQkFBS0NBUUVBL0hEY2w1Y0xQcUNVdDc5NElrS0Q0TzRPa0p6UXRweC9Qak1hZDMvNXNENTZxTEpkCnNMVDdtYXg4MzZoNnMzMkh3dkxRM1Fsb2V5KzcrbklxQ25KMHJRMzFSTmFhOENMWTVmakdsL05YTUIxMFJMUzIKalFDbWN5RTdiQlBrd3pzeXB0V0J0U1V3bnhEOEZCSzJqbndaTmNDd3BNMHAvSlR5aHBnRlhHREh2S3JJSDNLNgorZElmNjVJNWtsQXNwWEJxRHNFMlVULy9xbVlNMjgxNm9tWmIvbmlkUTA1ZmZETUZXdC9meDdqTFFNNHhyZkwzCkFDYnBDWVlXWko4ZTFwMGZjeEphamZrR3J6UCt2S1hEdXR1THBjd3gzZGw4aC9tQUQxL0ZJVzBkd3g0QjBlRlQKZy9YbmJLTmpqSVgxUllrOVplenovNUk5TWl1cEkyaittcGE0U3dJREFRQUJBb0lCQUQzTnIrUVpJYktpZytaOApwUC9XUXcxVm14RkNvMEtpQy9FQmQvQmJ6NWlibzFJYjRSNDdJUUlkbmJEd1FhSWlFSlBrK01lMDN2U1IyamJWCnJuOUNTRmg1KzF0d0l4eTVtMXFxVG95ekVkeUF5WTlzb3VqZ2lOL1UrZ1JhMW40Snh6dmJOOWRPbEpuTkFibzAKWHFlQnNIWmRoNkhCYlE4ZXRFWEpnRDUrbjBnTUYxcWI1N3NrMFYvM0lVMk1wVE50Q3h5Z0F5T0VjeVNEc253SgpENzhUTko0L1duUDBQZEtDTTFnMitqc0NLdXJUZDgvaTJvMUQwYlFUUkttNm9HZFF0OHhZMmordEpQOHdLUk5PCjVmMU9UWlR5bkt1ZnZkUlF6Q2tJUFFzSmRQYjh4bEtrb1VCNngxWmVvNWNxSVNrM3pyUHQxeGliTTdBa3lJWVYKdjcxWUhha0NnWUVBL2dtM2ppT0RhV3orcW5kTXFScFFKODB6aUMyOFo5cEFQWHNhNzZja1BkcmNMS0NsUzFXUwo5WHNNcUdha2xURXZkRHFQZWxRWTR1UExVZTYzbSswUm1rT3dmRzh6emZ5RzhwZzJTY2x2eHUvR0FFVVRoTk5VCm5CYTI3aEhzVFdTVTJXRFBEdFBqVm80Vnc5b3NZaTZhaUdLOERsanB6aUw0SW9EOGNIZElFVzBDZ1lFQS9tUDgKcGtscU9pamJmeE5qVkE5VGg2NmU1a3JWZ3hhREtkMFBGZ3RQVlY1WHBYRWtBQ1dxNWR1Ymx2TmRlbkhkbDlLcAplYU9EUzNFeW9HTFZWaHg0eVlEL2t4Yzd1YlN2RGJialVsNy9BUXlnSW9nSG9iVGpycXIvNDlvb1B6di9RTHdUCndQbjU1dlg1eU1TdjBxSWdIeElGUm4zUzh4VWhDZ3NCYTFxd2xaY0NnWUVBZ0RFSkZrbm5jMWpUL2ROcnROMloKV3l4YWQ3cVVtMWR3R1ZZNDFueGlkbkF0c0x0MERkcmw2TWdxTUFSend2RnAvOTNaTHlLbTJEK1pCd1JPVlBuQgphellESmhlN2JaT2p5V0Z6NjdJaG5LSzhLTUN5dzlFY25oTjhSaEIzVE9xWGprWWkxbG1vRzdKaFcwNktIUlZsClY3TnpNRTQvMVl5aDZFOXNrNWd0bnRVQ2dZRUF4V3Q3S3VGV213UVNNN0FzS1pRclV1VGxVL3RIYjVBTkFuUVgKdlJmUjNZSjFxbUwvZVUzQ0x0Z2JDeTZhb09lMC9KbjBBLzR0YW1MUzB6MWdLbysxL1VzckRFKy9mZ01HeHpIagpnSTFRZ05QM2dsVG05MHVXVm50R1V1RiswblFSUjMrdUpOamFZNjhJeGhJK0VNelkrTVBYYkRsQmVId3hROFgrCnQvMk5oTWtDZ1lFQXVTN2pOL1JMZUhGRzdQbEtmLzZrZXZjeEJ6QWFHb1J4dDdsSW9hQ0YxN01kQlNxMVdxS1IKVEpmWFk4aGMwUVpsRk1VUGluck9qRzVMNEVKdHRraG9UdHl2UVI3S3RkakhTUXlPMVNDZ1VwWTk1QzZvSXliMQp3ZGRaSjMwNVBZTFdmWkxEWXZ6bGU4aEhtUWtVQ3M0TklTUGJtWkJZYnZIVDl6SzFidkV3VTlrPQotLS0tLUVORCBSU0EgUFJJVkFURSBLRVktLS0tLQo=
  root-cert.pem: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUM1VENDQWMyZ0F3SUJBZ0lRRWdYaUg5VmVsYzVtUTJEY09CQjRwakFOQmdrcWhraUc5dzBCQVFzRkFEQWMKTVJvd0dBWURWUVFLRXhGck9ITXVZMngxYzNSbGNpNXNiMk5oYkRBZUZ3MHhPREEzTWpnd05qVTFNVEphRncweApPVEEzTWpnd05qVTFNVEphTUJ3eEdqQVlCZ05WQkFvVEVXczRjeTVqYkhWemRHVnlMbXh2WTJGc01JSUJJakFOCkJna3Foa2lHOXcwQkFRRUZBQU9DQVE4QU1JSUJDZ0tDQVFFQXN4NkJCSnFXclQzWXFrcERYUDdrdG5VYXpESW0KMnVmSjJIZ1JlV0ZwdzVrN0Y4NFV1TmdoYnNrdnlRZ2hpMkRlUXFHMzZrM0VoTkw0cGtGYU9ubXdIa1NFY2FNOAo4a1BGNXIrdTZRS2xvRVRwYzJBVytyUlRVRFdvS2dsdmhyL0IvUEVLN0FnZVQ3VXk1a3dTd2lBZ0w5OG1GVXVqCkhDUllNUGF5c2FtUVkxbXNreHhEK2tINS9YNHVNcU9DYVkrWDZDQnFtbkp3RmRUUENRcktaTVF2bjBDZXl6T3QKdHYvVWFSb1R3dWtNcnZvcFJmb2FoeE5uVlJXN0xaMWlhajZrWmhmRlR6SWpldTdPRDNUZElBbzUrc2h5Rm40SgpnUHYvaENQUmQ0aE9VSzhQRmZyTVlWQnZqNGxXN3N2VS9qcS9OTlpnMmxXMXdGZnhhd3FnTmhieUF3SURBUUFCCm95TXdJVEFPQmdOVkhROEJBZjhFQkFNQ0FnUXdEd1lEVlIwVEFRSC9CQVV3QXdFQi96QU5CZ2txaGtpRzl3MEIKQVFzRkFBT0NBUUVBbjU1S2hGTngzbE9EYmp4aTJVUVlCbjFRb3ZxdDN2ZUZjbWVTOFV2azMzaWxIS29ucVFHaQpZaHJPMEx5Vk1XYmV2V1h5OCsvMHlOVG8ySVozaWFyUUlaSk1qTkZUalhVRTY4b1N6N2ZqOW9XZUsyb2g2bE93CmdCUE9nVy9ZQ0FlK1g5RHRNVngrUUJMNlNyTjdWMmIrcVNWSEdXK0pOR3NoVFRsZzBHSW1QTDNMTEU2dU1vVWQKc0xZbFVxd2hlZExhY2VmVll6bFRxdHZNRy9XNk4yVkNDVVhvTS9BcUo4allYbXk3UlpnT1J1cEs2MVY5eWxHTQpWcmZuQ05PVnJMZWNlSVY1dWVCUG1KZkt0MUd0cnY4RmFoQlFxeWpyaVpQOXFNdWdFVC8vRzYzanlGeEdmbk5uCnVVVThlb2NaakxVQ0MxMFFzU0VNOURFaW9Ub3VpWmlORnc9PQotLS0tLUVORCBDRVJUSUZJQ0FURS0tLS0tCg==
kind: Secret
metadata:
  annotations:
    istio.io/service-account.name: istio-sidecar-injector
  creationTimestamp: 2018-07-28T09:47:43Z
  name: istio.istio-sidecar-injector
  namespace: istio-system
  resourceVersion: "32368"
  selfLink: /api/v1/namespaces/istio-system/secrets/istio.istio-sidecar-injector
  uid: 463fe694-924b-11e8-b5b0-0e46067ded37
type: istio.io/key-and-cert
```
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: istio-sidecar-injector
  namespace: istio-system
  labels:
    app: sidecar-injector
    istio: sidecar-injector
    release: {{ .Release.Name }}
    heritage: {{ .Release.Service }}
spec:
  replicas: {{ .Values.injector.replicaCount }}
  template:
    metadata:
      labels:
        app: sidecar-injector
        istio: sidecar-injector
        release: {{ .Release.Name }}
        heritage: {{ .Release.Service }}
    spec:
      containers:
        - name: sidecar-injector-webhook
          image: {{required "missing image" .Values.injector.image.name }}
          imagePullPolicy: {{ .Values.injector.image.pullPolicy }}
          args:
            - --caCertFile=/etc/istio/certs/root-cert.pem
            - --tlsCertFile=/etc/istio/certs/cert-chain.pem
            - --tlsKeyFile=/etc/istio/certs/key.pem
            - --injectConfig=/etc/istio/inject/config
            - --meshConfig=/etc/istio/config/mesh
            - --healthCheckInterval=2s
            - --healthCheckFile=/health
          volumeMounts:
          - name: config-volume
            mountPath: /etc/istio/config
            readOnly: true
          - name: certs
            mountPath: /etc/istio/certs
            readOnly: true
          - name: inject-config
            mountPath: /etc/istio/inject
            readOnly: true
          livenessProbe:
            exec:
              command:
                - /usr/local/bin/sidecar-injector
                - probe
                - --probe-path=/health
                - --interval=4s
            initialDelaySeconds: 4
            periodSeconds: 4
          readinessProbe:
            exec:
              command:
                - /usr/local/bin/sidecar-injector
                - probe
                - --probe-path=/health
                - --interval=4s
            initialDelaySeconds: 4
            periodSeconds: 4
          resources:
            requests:
              cpu: 10m 
      volumes:
      - name: config-volume
        configMap:
          name: istio
      - name: certs
        secret:
          secretName: istio.istio-sidecar-injector
      - name: inject-config
        configMap:
          name: istio-sidecar-injector
          items:
          - key: config
            path: config
      affinity:
        podAntiAffinity:
          preferredDuringSchedulingIgnoredDuringExecution:
          - weight: 1
            podAffinityTerm:
              topologyKey: kubernetes.io/hostname
              matchExpressions:
                labelSelector:
                  - key: app
                    operator: In
                    values:
                    - sidecar-injector
      serviceAccountName: istio-sidecar-injector
```

**Expected behavior**
Sidecar injector mutates the incoming pod creation request and mutated pods are deployed.

**Steps to reproduce the bug**
Apply webhook configuration and deploy sidecar injector with the usual cluster role bindings. May or may not be of note but I originally had the wrong cluster role binding so originally got permission denied. Have deleted everything and retried but with the same results.

**Version**

Istio images are all 1.0.0-snapshot.2

```
Liams-MBP:registry-charts liam$ k version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.2", GitCommit:"81753b10df112992bf51bbc2c2f85208aad78335", GitTreeState:"clean", BuildDate:"2018-05-12T04:12:12Z", GoVersion:"go1.9.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.5+IKS", GitCommit:"7593549b33fb8ab65a9a112387f2e0f464a1ae87", GitTreeState:"clean", BuildDate:"2018-07-19T06:26:20Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd6
```
```
Liams-MBP:registry-charts liam$ i version
Version: 4adeee62a473105c69326b8a17576df07f66b2ac
GitRevision: 4adeee62a473105c69326b8a17576df07f66b2ac
User: liam@Liams-MBP.connect
Hub: docker.io/istio
GolangVersion: go1.10.1
BuildStatus: Clean
```

**Is Istio Auth enabled or not?**
No

**Environment**
IBM Kubernetes Service

**Comment by cmluciano** [comment]
Cc @ayj

**Comment by ayj** [comment]
cc @ostromart 

Was this a transient or terminal error?

The validatingwebhookconfiguration's caBundle and citadel provisioned ca-cert.pem don't match. The sidecar injector should reconcile the former from the latter (~1 second). 

Can you check the injector logs? You should see an error "Patch webhook failed" with the reason if reconciliation fails. Can you also verify that the clusterrole and clusterrolebindings for sidecar injector are correct and that the injector pod is using them? You might need to restart the injector pod to adopt the new permissions if they were incorrect when the pod was first started.

**Comment by liamawhite** [comment]
Terminal

The ca-cert is the root-cert, right? I just did the following and still see the same error in the logs, no mention of webhook patch failed:

- Deleted the secret (citadel recreated)
- Deleted the pod (kube recreated)
- Set the caBundle in the webhook to be empty string (sidecar-injector repopulated it)

I also diffed root-cert.pem and caBundle, they are the same. Maybe this is a Kube issue? Do you think it tries to use the cert at the point at which the pod to be injected was scheduled?

**Comment by liamawhite** [comment]
OK, so I fixed this by reading another issue which I cannot find.

TL;DR the certs require the service account for the injector to not change from the chart default.

**Comment by Analyse4** [comment]
> OK, so I fixed this by reading another issue which I cannot find.
> 
> TL;DR the certs require the service account for the injector to not change from the chart default.

Hi, can you elaborate on the solution to this problem?I seem to have encountered the same situation as you.
